internal/provisioner/kops_provisioner.go

```diff
@@ -8,146 +8,105 @@ import (
 	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 
-	"github.com/mattermost/mattermost-cloud/internal/api"
 	"github.com/mattermost/mattermost-cloud/internal/model"
 	"github.com/mattermost/mattermost-cloud/internal/tools/kops"
 	"github.com/mattermost/mattermost-cloud/internal/tools/terraform"
 )
 
 // KopsProvisioner provisions clusters using kops+terraform.
 type KopsProvisioner struct {
-	clusterRootDir   string
-	s3StateStore     string
-	cs               clusterStore
-	kopsFactory      kopsFactoryFunc
-	terraformFactory terraformFactoryFunc
-	logger           log.FieldLogger
+	clusterRootDir string
+	s3StateStore   string
+	logger         log.FieldLogger
 }
 
 // NewKopsProvisioner creates a new KopsProvisioner.
-//
-// kopsFactory and terraformFactory exist purely to allow for unit testing. Passing nil defaults
-// to using the kops and terraform binary wrappers.
-func NewKopsProvisioner(clusterRootDir string, s3StateStore string, cs clusterStore, kopsFactory kopsFactoryFunc, terraformFactory terraformFactoryFunc, logger log.FieldLogger) *KopsProvisioner {
-	if kopsFactory == nil {
-		kopsFactory = func(logger log.FieldLogger) (KopsCmd, error) {
-			return kops.New(s3StateStore, logger)
-		}
-	}
-
-	if terraformFactory == nil {
-		terraformFactory = func(outputDir string, logger log.FieldLogger) TerraformCmd {
-			return terraform.New(outputDir, logger)
-		}
-	}
-
+func NewKopsProvisioner(clusterRootDir string, s3StateStore string, logger log.FieldLogger) *KopsProvisioner {
 	return &KopsProvisioner{
-		clusterRootDir:   clusterRootDir,
-		s3StateStore:     s3StateStore,
-		cs:               cs,
-		kopsFactory:      kopsFactory,
-		terraformFactory: terraformFactory,
-		logger:           logger,
+		clusterRootDir: clusterRootDir,
+		s3StateStore:   s3StateStore,
+		logger:         logger,
 	}
 }
 
 // CreateCluster creates a cluster using kops and terraform.
-func (provisioner *KopsProvisioner) CreateCluster(request *api.CreateClusterRequest) (*model.Cluster, error) {
-	provider, err := checkProvider(request.Provider)
-	if err != nil {
-		return nil, err
-	}
+func (provisioner *KopsProvisioner) CreateCluster(cluster *model.Cluster) error {
```

Comment:
No longer returning `*model.Cluster` -- just modify it directly as necessary.

Reply:
Why?

Reply:
The code used to interact with the store to create the cluster, but it's now created much earlier in the pipeline. I /could/ return the updated `*model.Cluster`, but it's weird to have to write:
```
return cluster, err
```

instead of the usual:
```
return nil, err
```

and if I do the latter, the caller has to avoid clobbering its existing `cluster` to record the updated error state. So I figured it would be easier to just let the parameter be modified directly.